smart_open/s3.py

```diff
@@ -279,47 +279,98 @@ class _SeekableRawReader(object):
     This class is internal to the S3 submodule.
     """
 
-    def __init__(self, s3_object, content_length, version_id=None, object_kwargs=None):
+    def __init__(self, s3_object, version_id=None, object_kwargs=None):
         self._object = s3_object
-        self._content_length = content_length
+        self._content_length = None
         self._version_id = version_id
         self._position = 0
         self._body = None
         self._object_kwargs = object_kwargs if object_kwargs else {}
 
-    def seek(self, position):
-        """Seek to the specified position (byte offset) in the S3 key.
+    def seek(self, offset, whence=constants.WHENCE_START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
 
-        :param int position: The byte offset from the beginning of the key.
+        Returns the position after seeking.
         """
+        if whence not in constants.WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % constants.WHENCE_CHOICES)
+
         #
         # Close old body explicitly.
         # When first seek() after __init__(), self._body is not exist.
         #
         if self._body is not None:
             self._body.close()
         self._body = None
-        self._position = position
 
-    def _load_body(self):
-        """Build a continuous connection with the remote peer starts from the current postion.
-        """
-        range_string = smart_open.utils.make_range_string(self._position)
-        logger.debug('content_length: %r range_string: %r', self._content_length, range_string)
+        start = None
+        stop = None
+        if whence == constants.WHENCE_START:
+            start = max(0, offset)
+        elif whence == constants.WHENCE_CURRENT:
+            start = max(0, offset + self._position)
+        else:
+            stop = max(0, -offset)
 
-        if self._position == self._content_length == 0 or self._position == self._content_length:
-            #
-            # When reading, we can't seek to the first byte of an empty file.
-            # Similarly, we can't seek past the last byte.  Do nothing here.
-            #
+        if self._content_length is not None and \
+                ((start is not None and start >= self._content_length) or stop == 0):
+            # avoid calling _load_body() if we are at the end of the file
             self._body = io.BytesIO()
+            self._position = self._content_length
         else:
-            self._body = _get(
+            range_string = smart_open.utils.make_range_string(start, stop)
+            self._load_body(range_string)  # sets self._position
+
+        return self._position
+
+    def _load_body(self, range_string=None):
+        """Build a continuous connection with the remote peer starts from the current postion.
```

Comment:
This docstring doesn't really match the name of the method.

If it were me, I would write something like:

> Reads the specified range of bytes over the wire. If no range is specified, reads the entire content. Stores the result in `_body`. Updates `_position` and `_content_length` as a side effect.

Reply:
True that the name and docstring don't match, but the docstring is more accurate -- it opens a connection but doesn't necessarily grab all the data yet. (Imagine calling `open(path).read(10)` on a gigabyte-large file.) I renamed the method to `_open_body` and expanded the docstring.